crypto/engine/eng_dyn.c

```diff
@@ -451,8 +451,17 @@ static int dynamic_load(ENGINE *e, dynamic_data_ctx *ctx)
          * We fail if the version checker veto'd the load *or* if it is
          * deferring to us (by returning its version) and we think it is too
          * old.
+         * Unfortunately the version checker does not distinguish between
+         * engines built for openssl 1.1.x and openssl 3.x, but loading
+         * an engine that is built for openssl 1.1.x will cause a fatal
+         * error.  Detect such engines, since EVP_PKEY_base_id is exported
+         * as a function in openssl 1.1.x, while it is a macro in openssl 3.x,
+         * and therefore only the symbol EVP_PKEY_get_base_id is available
+         * in openssl 3.x.
          */
-        if (vcheck_res < OSSL_DYNAMIC_OLDEST) {
+        if (vcheck_res < OSSL_DYNAMIC_OLDEST
+                || DSO_bind_func(ctx->dynamic_dso,
+                                 "EVP_PKEY_base_id") != NULL) {
```

Comment:
Having another look, doesn't this *reject* engines built against libcrypto 3.0?

Reply:
No, we have test cases, that would break if that was wrong.

Reply:
if this condition is found to be true, the engine is not loaded, as if it were a failed version check.
The symbol EVP_PKEY_base_id is only found in openssl 1.x, if this is built for openssl 1.x and
the symbol is not found for whatever reason, the worst thing that can happen, is that we continue,
as if we did not try to detect the situation.

Reply:
I did miss the different symbol name between here and #17541.  Thanks for pointing it out there.

That resolves this concern for me